File: js/NgramSuccessorModel.js
const Event = require('./Event.js').Event;

class NgramSuccessorModel {
    constructor() {
        this.successorMap = new Map();
        this.occurence = 0;
    }

    getProbability(event) {
        checkType(event);

        let successor = this.successorMap.get(event.key)
        if (successor == undefined)  {
            return 0;
        }
        return successor.occurence / this.occurence;

    }

    learn(event) {
        checkType(event);

        let successor = this.successorMap.get(event.key);
        if (successor == undefined) {
            successor = {
                event : event,
                occurence : 1
            }
        } else {
            successor.occurence++;
        }
        this.occurence++;
        this.successorMap.set(event.key, successor);
    }

}

function checkType(event) {
    if (event == null || event == undefined) {
        throw 'cannot learn null or undefined';
    }
    if (!(event instanceof Event)) {
        throw 'cannot learn event not Event';
    }
}

module.exports.NgramSuccessorModel = NgramSuccessorModel;

File: java/src/main/java/com/naturalness/NGramSuccessorModel.java
/*
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
*/

package com.naturalness;

import java.util.HashMap;
import java.util.Map;

public class NGramSuccessorModel {
    private final int NULL_PROBABILITY = 0;
    private NGram ngram;
    private Map<Event, Integer> successorMap;
    private int occurence;
    

    public NGramSuccessorModel(NGram ngram) {
        if (ngram == null) {
            throw new IllegalArgumentException("ngram cannot be null");
        }
        this.ngram = ngram;
        this.successorMap = new HashMap<Event, Integer>();
        this.occurence = 0;
    }

    public double getProbability(Event event) {
        if (event == null) {
            throw new IllegalArgumentException("event cannot be null");
        }
        if (!successorMap.containsKey(event)) {
            return NULL_PROBABILITY;
        }
        return successorMap.get(event).intValue() / occurence;
    }

    public void learn(Event event) {
        if (event == null) {
            throw new IllegalArgumentException("event cannot be null");
        }
        if (!successorMap.containsKey(event)) {
            successorMap.put(event, Integer.valueOf(1));
        } else {
            int newOccurence = successorMap.get(event).intValue()+1;
            successorMap.replace(event, Integer.valueOf(newOccurence));
        }
    }

}